contracts/stake/tests/common/prove.rs

```diff
@@ -0,0 +1,155 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Copyright (c) DUSK NETWORK. All rights reserved.
+
+use dusk_bytes::Serializable;
+use execution_core::{
+    plonk::{Circuit, Prover},
+    transfer::phoenix::{
+        InputNoteInfo, Prove, TxCircuit, TxCircuitVec, NOTES_TREE_DEPTH,
+    },
+};
+
+use rand::rngs::StdRng;
+use rand::SeedableRng;
+
+pub struct CachedProver();
+
+impl Prove for CachedProver {
+    fn prove(circuit: TxCircuitVec) -> Vec<u8> {
+        // fetch the prover from the cache and crate the circuit
+        match circuit.input_notes_info.len() {
+            1 => {
+                let (circuit, circuit_name) = tx_circuit_1_2(
+                    circuit
+                );
+                fetch_prover_and_prove(circuit, &circuit_name)
+            }
+            2 => {
+                let (circuit, circuit_name) = tx_circuit_2_2(
+                    circuit
+                );
+                fetch_prover_and_prove(circuit, &circuit_name)
+            }
+            3 => {
+                let (circuit, circuit_name) = tx_circuit_3_2(
+                    circuit
+                );
+                fetch_prover_and_prove(circuit, &circuit_name)
+            }
+            4 => {
+                let (circuit, circuit_name) = tx_circuit_4_2(
+                    circuit
+                );
+                fetch_prover_and_prove(circuit, &circuit_name)
+            }
+            _ => panic!("The `TxCircuit` is only implemented for 1, 2, 3 or 4 input-notes.")
+        }
```

Comment:
Most of the time if you have a function or a method with `_and_` in the name, is not correct: you're trying to make the function does two things. Also, you don't need to repeat the function call for each `match` branch. I would suggest to use the `match` as expression and returns the `(circuit, circuit_name)` and then split the function with the `_and_` (since the function we're in, is already called `prove`, I would perform the prove there).

Something like:

```rs
fn prove(circuit: TxCircuitVec) -> Vec<u8> {
  let (circuit, circuit_name) = match circuit.input_notes_info.len() {
            1 => tx_circuit_1_2(circuit),
            2 => tx_circuit_2_2(circuit),
            3 => tx_circuit_3_2(circuit),
            4 => tx_circuit_4_2(circuit),
            _ => panic!("The `TxCircuit` is only implemented for 1, 2, 3 or 4 input-notes.")
  };

  let prover = fetch_prover(circuit, &circuit_name);

  let (proof, _pi) = prover
        .prove(rng, &circuit)
        .expect("The circuit should be correct");

    proof.to_bytes().into()
  }
}
```

Also, are we sure we want always to panic, and never return `Result` to handle the errors? In general, `tests` can returns `Result` and be handled by the test harness well.

Reply:
Good one, I'll use the lazy prover as we do in `rusk-prover`